Step2 : 문자열 계산기

jpoh281 opened the pull request:
안녕하세요? 2단계 완료 리뷰요청드립니다.

실제 계산기와 비슷한 느낌으로 만들고 싶어서 생각하다 보니까 아래와 같이 클래스를 나눠서 구현했습니다.
Calculator - 계산기
Button - 입력
Monitor - 출력
Memory - 입력받은 값과, 누산된 값을 기억
CPU - 연산

- 덧셈
- 뺄셈
- 곱셈
- 나눗셈
- 입력값이 null이거나 빈 공백 문자일 경우 IllegalArgumentException throw
- 사칙연산 기호가 아닌 경우 IllegalArgumentException throw
- 사칙 연산을 모두 포함하는 기능 구현

질문
1. 커밋 메시지를 이렇게 작성해보지 않아서 되게 어색했어요. 혹시라도 더 자연스럽게 변경할 수 있는 부분이 있다면 알려주시면 감사하겠습니다!

2. CPU 테스트를 만드는 과정에서 `3 / 2 = 1` 이라고 함수명을 정하고 싶었는데 '/'가 사용할 수 없는 기호여서 Kotest를 사용했어요.
Kotest 정말 좋은 것 같습니다. 이럴 때 만약 리뷰어님은 어떤 메세지를 사용하셨을지도 궁금해요.

3. 버튼 테스트를 위해 readLine을 사용하는 메소드와, 파라미터로 넘겨받는 메소드를 구현했는데 이런식으로 하는게 맞을까요?

Commit:
feat: 계산기 사칙연산 구현

jpoh281 committed:
feat: Calculator의 사칙 기능을 CPU로 이동

jpoh281 committed:
refactor: CalculatorTest를 CPUTest로 이동

jpoh281 committed:
feat: 입력을 담당할 Button, 출력을 담당할 Monitor 추가

jpoh281 committed:
feat: Button Class 구현

jpoh281 committed:
feat: Monitor Class 구현

jpoh281 committed:
study: isDigit 테스트

Commit:
feat: 버튼 클래스 테스트 메소드 추가

Commit:
feat: 메모리 클래스 및 테스트 생성

Commit:
feat: 최종 계산기 구현 완료

BeokBeok commented on src/main/kotlin/calculator/Button.kt:
```diff
@@ -0,0 +1,21 @@
+package calculator
+
+class Button {
```
입력하는 클래스 이름을 Button 이라고 작성해주셨네요
다만, 저는 Button을 보고서 "입력"과 관련된 기능을 하는지 잘 와닿지 않네요. 😅 
입력이라고 한다면, 입력의 의미에 맞게 클래스 이름을 변경해보면 어떨까요?

BeokBeok commented on src/main/kotlin/calculator/Button.kt:
```diff
@@ -0,0 +1,21 @@
+package calculator
+
+class Button {
+    fun pressed(): String {
+        val input: String? = readLine()
+
+        if (input.isNullOrBlank()) {
+            throw IllegalArgumentException()
+        }
+
+        return input
+    }
+
+    fun pressed(input: String?): String {
+        if (input.isNullOrBlank()) {
+            throw IllegalArgumentException()
+        }
+
+        return input
+    }
```
pressed라는 함수를 오버로딩하여 두 개를 만들어주셨네요. 👍 
다만, `val input: String? = readLine()` 이 코드를 제외하고는 전부 동일하네요.
수업시간에 배우셨듯이, 코틀린에는 기본값을 지정할 수 있습니다.
기본값을 활용해서 중복된 코드를 줄여보면 어떨까요?

BeokBeok commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,63 @@
+    private var memory: Memory = Memory()
+) {
+
+    init {
+        print("----on----")
+    }
```
Calculator 클래스에서 출력과 관련된 동작을 하고 있네요.
출력의 책임을 갖는 Monitor 클래스에서 출력하도록 코드를 개선해보면 어떨까요?

BeokBeok commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,63 @@
+        return expression.split(" ")
+    }
+
+    private fun parse(term: String, operand1: Int, operand2: Int): Int {
+        return when (term) {
+            "+" -> {
+                cpu.add(operand1, operand2)
+            }
+
+            "-" -> {
+                cpu.subtract(operand1, operand2)
+            }
+
+            "*" -> {
+                cpu.multiply(operand1, operand2)
+            }
+
+            "/" -> {
+                cpu.divide(operand1, operand2)
+            }
+
+            else -> throw IllegalArgumentException()
+        }
```
사칙연산 기호에 맞는 동작을 구현해주셨네요. 👍 
만약, %와 같이 나눈 나머지를 구하는 연산이 추가되면 이 코드는 어떻게 동작하게 될까요?
연관된 상수값들은 enum을 활용해서 코드를 개선해보면 어떨까요?

ref. https://kotlinlang.org/docs/enum-classes.html#implementing-interfaces-in-enum-classes

BeokBeok commented on src/main/kotlin/calculator/Memory.kt:
```diff
@@ -0,0 +1,40 @@
+package calculator
+
+class Memory(private var terms: List<String> = emptyList(), var accumulator: Int = 0) {
```
accumulator를 외부에서 값을 변경하도록 하고 있네요. 🤔 
지금은 Memory와 Calculator에서 값을 변경하고 있지만, 수많은 클래스에서 accumulator에 접근해서 값을 변경하게 된다면 어떤 시점에 어디에서 변경했는지 알기가 어려워요.
val을 활용해서 외부에서 값을 변경하지 못하도록 개선해보면 어떨까요?

BeokBeok commented on src/main/kotlin/calculator/Memory.kt:
```diff
@@ -0,0 +1,40 @@
+        accumulator = 0
+    }
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (javaClass != other?.javaClass) return false
+
+        other as Memory
+
+        if (terms != other.terms) return false
+        if (accumulator != other.accumulator) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
+        var result = terms.hashCode()
+        result = 31 * result + accumulator
+        return result
+    }
```
코틀린에서는 data class라는 것을 제공해요.
data class는 이와 같이 equals, hashCode, toString등 같은 값을 가진 경우에는 같은 객체로 판단하도록 내부에 구현이 되어 있어요.
data class를 활용해보면 어떨까요?

BeokBeok commented on src/test/kotlin/calculator/CPUTest.kt:
```diff
@@ -0,0 +1,35 @@
+        CPU().divide(4, 2) shouldBe 2
+    }
+
+    // 함수명에 /(Slash)를 넣고 싶은데 안되서 JUnit말고 kotest를 사용했습니다.
+    "4 / 0 = ArithmeticException(\"/ by zero\")" {
+        shouldThrowWithMessage<ArithmeticException>("/ by zero") {
+            CPU().divide(4, 0)
+        }
+    }
```
함수명에 "/"를 넣어야 하는 이유가 있을까요?
팀원들이 봤을 때, 이 함수명만 보고서 어떤 것을 테스트하는지 알 수 있을까요? 🤔 
테스트 코드에 함수명은 주로 수식을 적기 보다는, given-when-then 방식으로 적어줍니다.
"0으로 나누면_예외가 발생한다" 와 같이 작성하는 것이 읽기가 편합니다.

BeokBeok commented on src/test/kotlin/calculator/MemoryTest.kt:
```diff
@@ -0,0 +1,41 @@
+
+class MemoryTest {
+    @Test
+    fun `기본 인자`() {
```
기본 인자 라는 것은 어떤 것을 의미할까요?
테스트하고자 하는 것을 좀 더 명시적으로 작성해보면 어떨까요? (이하 관련 내용 동일)

BeokBeok commented on src/test/kotlin/calculator/study/IsDigitTest.kt:
```diff
@@ -0,0 +1,17 @@
+class IsDigitTest {
+
+    @Test
+    fun `1 is Digit`() {
```
테스트 함수명을 한글, 수식, 영어 등 혼합해서 사용하고 있네요.
테스트 함수명을 한 가지 양식으로 통일해보면 어떨까요?

BeokBeok reviewed (changes requested):
문자열 계산기 미션하시느라 고생하셨습니다. 👍 
고민해볼만한 의견들을 코맨트로 작성하였으니, 충분히 고민해보시고 도전해보세요. 💪

BeokBeok commented:
> 커밋 메시지를 이렇게 작성해보지 않아서 되게 어색했어요. 혹시라도 더 자연스럽게 변경할 수 있는 부분이 있다면 알려주시면 감사하겠습니다!

A. 모든지 처음은 어색하기 마련입니다. 😅 
지금도 커밋 메시지를 잘 작성해주시고 계십니다. 💪 
계속 하시다보면 익숙해지고, 나아가서는 요령이 생길 것 같아요!

BeokBeok commented:
> CPU 테스트를 만드는 과정에서 3 / 2 = 1 이라고 함수명을 정하고 싶었는데 '/'가 사용할 수 없는 기호여서 Kotest를 사용했어요.
Kotest 정말 좋은 것 같습니다. 이럴 때 만약 리뷰어님은 어떤 메세지를 사용하셨을지도 궁금해요.

A. 코맨트에도 작성했듯이, 테스트 함수명은 given-when-then 방식으로 적는 것을 권장합니다.
그래야 어떤 것을 테스트하려고 하는지 함수명만 보고도 파악하기가 쉽습니다. 😄

BeokBeok commented:
> 버튼 테스트를 위해 readLine을 사용하는 메소드와, 파라미터로 넘겨받는 메소드를 구현했는데 이런식으로 하는게 맞을까요?

A. https://github.com/next-step/kotlin-racingcar/pull/738#discussion_r1013973750 코맨트 작성했습니다 😄

jpoh281 committed:
refactor: Button 클래스를 Keypad로 변경

https://github.com/next-step/kotlin-racingcar/pull/738#discussion_r1013971742
해당 피드백을 보고 Button보다 좀 더 와닿을 것 같은 이름으로 해봤습니다.

jpoh281 committed:
refactor: Calculator 내부의 print를 Monitor 가 하도록 변경

https://github.com/next-step/kotlin-racingcar/pull/738#discussion_r1013977334

jpoh281 committed:
refactor: Memory class를 data class로 변경

https://github.com/next-step/kotlin-racingcar/pull/738#discussion_r1013983451

jpoh281 committed:
refactor: 외부에서 누산기의 값을 수정하지 못하도록 변경, 테스트 수정

https://github.com/next-step/kotlin-racingcar/pull/738#discussion_r1013981922
copy를 이용해서 불변객체를 만들어야하는지, getter를 이용해서 외부에서만
수정 불가능하게 하는것이 좋을지 생각해봤는데 일단은 getter를 통해서 외부에서만
값을 변경시키지 못하게 하는것이 좋다는 생각이 들었습니다.

jpoh281 committed:
refactor: 기본 인자를 이용해서 메소드 하나로 변경

https://github.com/next-step/kotlin-racingcar/pull/738#discussion_r1013973750

jpoh281 committed:
feat: 연산자 기호 추가 및 테스트 추가

https://github.com/next-step/kotlin-racingcar/pull/738#discussion_r1013978637

이넘을 만들었는데 어떻게 찾아야할지 몰라서 다른분들의 깃허브를 살펴봤습니다.

jpoh281 committed:
refactor: 테스트 명칭 한글로 통일성있게 수정

https://github.com/next-step/kotlin-racingcar/pull/738#discussion_r1013990681

jpoh281 commented:
좋은 리뷰 해주셔서 감사합니다. 재미있는 피드백들이어서 즐겁게 코드 수정했어요. 😄
커밋들에 코멘트 링크 달았는데 보실 때 도움이 되셨으면 좋겠어요.

작성하면서 궁금했던 부분을 또 여쭤보자면
1. Button -> Keypad 로 변경했는데 이게 실제 계산기에서 찾지말고 진짜 Input과 관련된 무언가로 바꾸라는 말씀이셨을까요..?
마땅히 떠오르는 클래스 명이 잘 없는 것 같아요.

2. accumulator를 val로 해서 외부에서 수정 못하게 하라고 피드백해주셨는데
이게 불변객체를 만들라는 뜻인지 내부에서만 수정가능하도록 하는걸 추천해주셨는지 애매했어요. 
저는 일단 외부에서만 값을 수정시키지 못하도록 했습니다.

감사합니다.

BeokBeok commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,61 @@
+        return expression.split(" ")
+    }
+
+    private fun parse(operator: Operator, operand1: Int, operand2: Int): Int {
+        return when (operator) {
+            Operator.ADD -> {
+                cpu.add(operand1, operand2)
+            }
+
+            Operator.SUBTRACT -> {
+                cpu.subtract(operand1, operand2)
+            }
+
+            Operator.MULTIPLY -> {
+                cpu.multiply(operand1, operand2)
+            }
+
+            Operator.DIVIDE -> {
+                cpu.divide(operand1, operand2)
+            }
+        }
+    }
```
Caluclator 클래스에서 사칙연산에 따라 연산을 처리해주고 있네요.
하지만 Caluclator 클래스는 계산 외에도 Operator를 parse하는 책임도 가지고 있네요. 🤔 
Calculator 클래스는 하나의 책임 즉, "계산" 에 대한 책임만 가져야합니다.
Operator에 관련된 책임은 Operator 객체가 갖도록 변경해보면 어떨까요?

BeokBeok commented on src/main/kotlin/calculator/Operator.kt:
```diff
@@ -0,0 +1,20 @@
+
+    companion object {
+        fun findByValue(prime: String): Operator {
+            Operator.values()
+                .forEach { operator ->
+                    if (prime == operator.prime) {
+                        return operator
+                    }
+                }
```
```suggestion
            values().find { operator ->
                prime == operator.prime
            }
```
Operator에 정의된 연산자인지 아닌지 찾는 함수를 잘 구현해주셨네요. 👍 
정의된 연산자를 찾기 위해 forEach를 활용해주셨는데요.
forEach를 활용해주셔도 좋지만, find 함수를 사용하면 좀 더 간결하게 코드를 작성하실 수 있어요. 😄

BeokBeok commented on src/test/kotlin/calculator/CPUTest.kt:
```diff
@@ -0,0 +1,35 @@
+        CPU().divide(4, 2) shouldBe 2
+    }
+
+    // 함수명에 /(Slash)를 넣고 싶은데 안되서 JUnit말고 kotest를 사용했습니다.
```
```suggestion
```
불필요한 주석은 제거해보면 어떨까요?

BeokBeok reviewed (approved):
피드백 반영하시느라 고생하셨습니다. 👍 
추가로 고민해볼만한 의견들을 코맨트로 작성하였으니, 충분히 고민해보시고 도전해보세요. 💪 
다음 미션을 진행하시면서, 추가로 작성된 코맨트들도 반영해주세요. 🙏

BeokBeok commented:
> Button -> Keypad 로 변경했는데 이게 실제 계산기에서 찾지말고 진짜 Input과 관련된 무언가로 바꾸라는 말씀이셨을까요..?
마땅히 떠오르는 클래스 명이 잘 없는 것 같아요.

A. 실제 계산기에서 찾기 보다는 Input과 관련된 무언가로 바꾸는 것을 의도했습니다. 😅 
그래도 Button 보다는 Keypad가 더 이해하기 좋은 것 같습니다. 👍

BeokBeok commented:
> accumulator를 val로 해서 외부에서 수정 못하게 하라고 피드백해주셨는데
이게 불변객체를 만들라는 뜻인지 내부에서만 수정가능하도록 하는걸 추천해주셨는지 애매했어요.
저는 일단 외부에서만 값을 수정시키지 못하도록 했습니다.

A. 아마 다음 미션에서 불변객체에 대해서 학습하시게 되어서 관련해서 언급을 드리지 않았습니다. 😅 
제가 의도한 것은, 외부에서 직접 접근해서 값을 변경하지 않도록 하는 것이였어요! 👍
